compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -6957,6 +6957,16 @@ private BType checkObjectFieldAccessExpr(BLangFieldBasedAccess fieldAccessExpr,
 
     private BType checkRecordFieldAccessExpr(BLangFieldBasedAccess fieldAccessExpr, BType varRefType, Name fieldName) {
         if (varRefType.tag == TypeTags.RECORD) {
+            BSymbol fieldSymbol = symResolver.resolveStructField(fieldAccessExpr.pos, this.env,
+                    fieldName, varRefType.tsymbol);
+
+            if (Symbols.isOptional(fieldSymbol) && !fieldSymbol.type.isNullable() && !fieldAccessExpr.isLValue) {
+                fieldAccessExpr.symbol = fieldSymbol;
+                return addNilForNillableAccessType(fieldSymbol.type);
+            } else if (fieldAccessExpr.expr.getBType().tag == TypeTags.UNION &&
+                    isFieldOptionalInRecords(fieldAccessExpr, fieldName) && fieldSymbol.type.isNullable()) {
```

Comment:
Do we need to do this? Looking back to the field access expression?

If the field is 1. optional  2. of a nilable type and 3. !fieldAccessExpr.isLValue (only if we need to consider isLValue here) we can return `symTable.semanticError` here, right?

Then wouldn't the check at L6982 handle returning `symTable.semanticError`?